### DISCUSSION TITLE
Supabase & Vercel Environments Mapping

### DISCUSSION THREAD

**Discussion by dmudro** [supabase/supabase#11888] [Questions]
following up on discussion in [Multiple environments and migrations – best practices?](https://github.com/supabase/supabase/discussions/542). I believe to have a pretty common stack / scenario:
- Vercel + Next.js + Supabase
- need completely separate staging-like AND completely separate dev-like environments for Supabase
- these should match Vercel's development / preview / production
- besides preview per feature branch, I have a 'permanent' staging git branch mapped to a custom staging domain / deployment in Vercel but under the hood it's just plain Vercel preview env

my current solution is:
1. local Supabase container based instance
2. linked to a (free tier) supabase 'preview' project (so I can `supabase diff`)
3. completely separate pro supabase 'production' project not linked to anything

is this a valid envs strategy these days given the common stack above?

Ideally I would have avoided 2 different projects that need to be synced manually for schema changes etc. a perfect solution would be some sort of branching or shadow / private / sandbox instances *within one Supabase project* which does not seem to be the case in the current product.

**Comment by dmudro** [comment]
I just stumbled upon a note about multi-environment project in [pricing FAQ](https://supabase.com/pricing).

> We are working on multi-environment projects. For now, you can create a project for your development backend and production backend.

**Reply by einarpersson** [reply]
I was just about to write about the same thing! I wrote about it in Discord (but got no feedback)

On every new branch you get a vercel deployment. I would like to run E2E-tests against that deployment (as I do today) but then I also need a temporary database per branch.

##  Message on Discord:
Hi!

I use SvelteKit deployed on Vercel. 

The Vercel-github-integration automatically creates a Vercel deployment on each PR which is great. I can both manually inspect and run E2E-tests using Playwright. When all checks have run I can safely merge into main which deploys the site. This workflow is very lightweight and works well for my purposes.

**What I would like is for Supabase to mirror the behaviour of Vercel, i.e. create a temporary db for each new PR** (using migrations in the branch) so that I run E2E-tests against a deployed vercel preview version *and* a deployed "supabase preview version". 

I suspect this behaviour is not supported currently, but I wonder if it is on the radar? Vercel PR preview deployments are great, but they become close to useless (?) without a guaranteed to be in sync with any migrations in the PR. 

I saw this 
> *Developers should be able to "fork" a database at any point. This is particularly important for Jamstack developers who run deployments for every Git branch.*
from https://supabase.com/blog/supabase-series-b 

Maybe I misunderstand this, but it sounds close to what I am after?

--

Summary:
1. Is this supported?
2. If not, is this being worked on or planned?
3. If not, do you see a way of achieving this on my own? Eg. could I use the CLI to create a new temporary project for each PR (in a github workflow) and delete it on close?

**Reply by einarpersson** [reply]
To add to the confusion, I see that there is some kind of branching support in the CLI but it is not documented on the official page, nor does it seem to support cloud/remote hosted db.

[found it here](https://fig.io/manual/supabase/db/branch)
[supabase docs](https://supabase.com/docs/reference/cli/supabase-db) (not here!)

![image](https://user-images.githubusercontent.com/6043691/215979511-eaf4ab00-b8b5-4e78-943f-91e1a53367a5.png)

**Reply by dmudro** [reply]
solid env / db branching instead of separate project for preview env... great candidate for a future launch week! 🙏
